### PR TITLE
update: don't generate useless Into<impl T>

### PR DESCRIPTION
**Pull request by Velfi**
## Motivation and Context

[smithy#990](https://github.com/awslabs/smithy-rs/issues/990)

## Description

Stop the code generator from inserting `impl Into` where it's not helpful

## Testing

Tested by running existing tests

## Checklist

- [x] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates
- [x] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/BuilderGenerator.kt [hunk 1]
```diff
@@ -217,17 +223,24 @@ class BuilderGenerator(
...
+        val k = when (coreType.key) {
+            is RustType.String,
+            is RustType.Box -> "impl Into<${coreType.key.render(true)}>"
+            else -> "${coreType.key.render(true)}"
+        }
+        val v = when (coreType.member) {
+            is RustType.String,
+            is RustType.Box -> "impl Into<${coreType.member.render(true)}>"
+            else -> "${coreType.member.render(true)}"
+        }
```
Comment (rcoh) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/BuilderGenerator.kt:235]: lets extract this snippet into a function

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/BuilderGenerator.kt [hunk 2]
```diff
@@ -217,17 +223,24 @@ class BuilderGenerator(
...
-            coreType.member.render(
-                true
-            )
-            }>) -> Self"
+            "pub fn $memberName(mut self, k: $k, v: $v) -> Self"
         ) {
             rust(
                 """
                 let mut hash_map = self.$memberName.unwrap_or_default();
-                hash_map.insert(k.into(), v.into());
```
Comment (rcoh) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/BuilderGenerator.kt:230]: probably need to conditionally generate the `.into()` invocations here

Comment (Velfi) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/BuilderGenerator.kt:230]: yeah, based on this issue and the thing above I'll make a util that returns a tuple of the type and the value ie `["impl Into<${coreType.member.render(true)}>", "v.into()"]`

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/BuilderGenerator.kt (hunk 1 above)
Comment (Velfi) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/BuilderGenerator.kt]: I extracted the snippet but I feel like I did it in a very caveman manner and I'm open to suggestions if you think it's gross. I wasn't sure where they belonged and I also struggled to come up with informational names.

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/FluentClientDecorator.kt [hunk 3]
```diff
@@ -472,14 +473,11 @@ class FluentClientGenerator(
                             is RustType.Vec -> with(core) { renderVecHelper(member, memberName, coreType) }
                             is RustType.HashMap -> with(core) { renderMapHelper(member, memberName, coreType) }
                             else -> {
-                                val signature = when (coreType) {
-                                    is RustType.String,
-                                    is RustType.Box -> "(mut self, inp: impl Into<${coreType.render(true)}>) -> Self"
-                                    else -> "(mut self, inp: ${coreType.render(true)}) -> Self"
-                                }
+                                val signature = coreType.asArgument("signature")
```
Comment (rcoh) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/FluentClientDecorator.kt:476]: did you mean to name this `signature`?

**Review by rcoh — APPROVED** [review]
LGTM! this is great, love seeing old code get cleaned up

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/FluentClientCore.kt [hunk 4]
```diff
@@ -38,13 +39,13 @@ class FluentClientCore(private val model: Model) {
         docs("To override the contents of this collection use [`${member.setterName()}`](Self::${member.setterName()}).")
         rust("///")
         documentShape(member, model)
-        val k = coreType.key
-        val v = coreType.member
+        val (k_param, k_usage) = coreType.formattedKey()
```
[the remark this answers is not in the document]
Comment (Velfi) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/FluentClientCore.kt]: huh this is outdated code, it actually looks like this:
```
        val k = coreType.key.asArgument("k")
        val v = coreType.member.asArgument("v")
```

**Pull request merged**